File: fastapi-service/gemini.py
"""
Gemini AI integration for resume enhancement and ATS analysis
"""

import google.generativeai as genai
import os
from typing import Dict, List, Optional
import json
import logging

logger = logging.getLogger(__name__)

# Direct API key configuration (temporary for testing)
api_key = "AIzaSyAB6SXYVqMItI7Zk44GI1WoS8vaLgNY8do"
os.environ["GEMINI_API_KEY"] = api_key
# Configure Gemini API
try:
    genai.configure(api_key=api_key)
    logger.info("Successfully configured Gemini API")
except Exception as e:
    logger.error(f"Error configuring Gemini API: {str(e)}")
    raise ValueError(f"Failed to configure Gemini API: {str(e)}")

# Configure Gemini API
genai.configure(api_key=api_key)

def get_gemini_model():
    """Get the most capable available Gemini model"""
    try:
        model = genai.GenerativeModel('gemini-1.5-flash')
        return model
    except Exception as e:
        logger.error(f"Error getting Gemini model: {str(e)}")
        return None

def format_resume_sections(text: str) -> Dict:
    """
    Use Gemini to format and structure the resume text into clear sections
    """
    try:
        model = get_gemini_model()
        if not model:
            raise Exception("No suitable Gemini model found")

        prompt = f"""
        You are a professional resume parser and formatter. Format and structure the following resume text into clear sections.
        Identify and organize the following sections if present:
        - Personal Information (name, contact, etc.)
        - Summary/Objective
        - Skills
        - Work Experience
        - Education
        - Projects
        - Certifications
        - Additional Information

        For each work experience and education entry, ensure to extract and format:
        - Company/Institution name
        - Position/Degree
        - Dates
        - Key responsibilities/achievements

        Return the response as a structured JSON with these sections as keys.
        
        Resume text to parse:
        {text}
        """

        response = model.generate_content(prompt)
        if response and response.text:
            # Try to parse the response as JSON
            try:
                # Clean up the response text to ensure it's valid JSON
                json_str = response.text.strip()
                if json_str.startswith("```json"):
                    json_str = json_str.split("```json")[1]
                if json_str.endswith("```"):
                    json_str = json_str[:-3]
                
                # Log the raw and cleaned response
                logger.info("Raw Gemini response: %s", response.text)
                logger.info("Cleaned JSON string: %s", json_str)
                
                formatted_sections = json.loads(json_str.strip())
                logger.info("Parsed JSON response: %s", json.dumps(formatted_sections, indent=2))
                return formatted_sections
            except json.JSONDecodeError as e:
                logger.error(f"Error parsing Gemini response as JSON: {str(e)}")
                # Return a basic structure with the raw text
                return {"raw_text": response.text}
        else:
            raise Exception("No response from Gemini")

    except Exception as e:
        logger.error(f"Error in format_resume_sections: {str(e)}")
        return {"error": str(e), "raw_text": text}

def enhance_resume_with_ai(resume_text: str) -> Dict:
    """
    Enhance and format resume text using Gemini AI
    """
    try:
        model = get_gemini_model()
        if not model:
            raise Exception("No suitable Gemini model found")

        # First, format the resume into sections
        formatted_sections = format_resume_sections(resume_text)

        # Then, enhance each section
        prompt = f"""
        You are a professional resume enhancer. Improve the following resume sections to make them more impactful and ATS-friendly.
        For each bullet point:
        1. Use strong action verbs
        2. Include measurable achievements
        3. Remove fluff and passive language
        4. Ensure relevant keywords are included
        5. Make achievements quantifiable where possible

        Resume sections to enhance:
        {json.dumps(formatted_sections, indent=2)}
        
        Return the enhanced version in the same JSON structure.
        """

        response = model.generate_content(prompt)
        if response and response.text:
            try:
                # Clean up the response text to ensure it's valid JSON
                json_str = response.text.strip()
                if json_str.startswith("```json"):
                    json_str = json_str.split("```json")[1]
                if json_str.endswith("```"):
                    json_str = json_str[:-3]
                
                # Log the raw and cleaned response for enhancement
                logger.info("Raw enhancement response: %s", response.text)
                logger.info("Cleaned enhancement JSON string: %s", json_str)
                
                enhanced_sections = json.loads(json_str.strip())
                logger.info("Parsed enhancement JSON: %s", json.dumps(enhanced_sections, indent=2))
                return enhanced_sections
            except json.JSONDecodeError as e:
                logger.error(f"Error parsing enhanced resume JSON: {str(e)}")
                return formatted_sections
        else:
            return formatted_sections

    except Exception as e:
        logger.error(f"Error in enhance_resume_with_ai: {str(e)}")
        return {"error": str(e), "original_text": resume_text}

def analyze_ats_score(resume_text: str, job_description: str) -> Dict:
    """
    Analyze ATS compatibility using Gemini AI
    """
    try:
        model = get_gemini_model()
        if not model:
            raise Exception("No suitable Gemini model found")

        prompt = f"""
        Analyze this resume against the job description for ATS compatibility.
        
        Resume:
        {resume_text}
        
        Job Description:
        {job_description}
        
        Provide a detailed analysis in JSON format with the following structure:
        {
            "score": <overall score 0-100>,
            "analysis": {
                "keyword_match": <score 0-100>,
                "format_compatibility": <score 0-100>,
                "missing_keywords": [list of important missing keywords],
                "suggestions": [list of specific improvements]
            }
        }
        """

        response = model.generate_content(prompt)
        if response and response.text:
            try:
                # Clean up and parse the response
                json_str = response.text.strip()
                if json_str.startswith("```json"):
                    json_str = json_str.split("```json")[1]
                if json_str.endswith("```"):
                    json_str = json_str[:-3]
                
                # Log the responses
                logger.info("Raw ATS analysis response: %s", response.text)
                logger.info("Cleaned ATS analysis JSON string: %s", json_str)
                
                analysis_result = json.loads(json_str.strip())
                logger.info("Parsed ATS analysis: %s", json.dumps(analysis_result, indent=2))
                return analysis_result
            except json.JSONDecodeError as e:
                logger.error(f"Error parsing ATS analysis JSON: {str(e)}")
                return {
                    "score": 75,
                    "error": str(e),
                    "analysis": {
                        "keyword_match": 70,
                        "format_compatibility": 80,
                        "missing_keywords": [],
                        "suggestions": ["Error parsing AI response - using fallback values"]
                    }
                }
    except Exception as e:
        logger.error(f"Error in ATS analysis: {str(e)}")
        return {
            "score": 75,
            "error": str(e),
            "analysis": {
                "keyword_match": 70,
                "format_compatibility": 80,
                "missing_keywords": [],
                "suggestions": ["Error during analysis - using fallback values"]
            }
        }

def extract_skills_from_text(text: str) -> List[str]:
    """
    Extract skills from text using AI
    TODO: Implement skill extraction
    """
    # Placeholder implementation
    return ["Python", "JavaScript", "React", "Node.js"]

File: lib/ai/gemini.ts
import { GoogleGenerativeAI } from "@google/generative-ai"

const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY!)

export async function parseResumeWithAI(resumeText: string) {
  try {
    console.log("\n=== Starting Resume Parsing ===")
    const model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" })

    const prompt = `Format the following resume text as structured JSON data.
    The response must be ONLY valid JSON with these exact fields:
    {
      "name": "Full Name",
      "email": "email@example.com", 
      "phone": "phone number",
      "summary": "professional summary",
      "skills": ["skill1", "skill2", "skill3"],
      "experience": "work experience details",
      "education": "education details", 
      "certifications": ["cert1", "cert2"],
      "projects": ["project1", "project2"]
    }

    Resume text to parse:
    ${resumeText}

    Remember: Return ONLY the JSON object with no other text or markdown formatting.`

    console.log("\n=== Parsing Resume with Gemini ===")
    console.log("Sending prompt to Gemini:")
    console.log(prompt)
    console.log("================================\n")

    const result = await model.generateContent(prompt)
    const response = await result.response
    
    // Parse JSON directly from the response
    try {
      const text = response.text()
      if (!text) throw new Error("Empty response from AI")
      
      console.log("\nRaw response from Gemini:")
      console.log("------------------------")
      console.log(text)
      console.log("------------------------")
      
      // First try direct JSON parse
      try {
        return JSON.parse(text)
      } catch {
        // If direct parse fails, try to extract JSON
        const jsonMatch = text.match(/\{[\s\S]*\}/)
        if (jsonMatch) {
          return JSON.parse(jsonMatch[0])
        }
        throw new Error("No valid JSON found in response")
      }
    } catch (parseError) {
      console.error("JSON parsing error:", parseError)
      throw new Error("Failed to parse AI response into JSON")
    }

    throw new Error("Failed to parse JSON from AI response")
  } catch (error) {
    console.error("Error parsing resume with AI:", error)
    throw error
  }
}

export async function enhanceResumeWithAI(resumeData: any, jobDescription?: string) {
  try {
    const model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" })

    const prompt = jobDescription
      ? `You are an expert resume writer. Enhance this resume for the given job description.
        Return ONLY a valid JSON object with the enhanced content, keeping the exact same structure.
        Do not include any other text or markdown formatting in your response.
        
        Job Description:
        ${jobDescription}
        
        Resume Data:
        ${JSON.stringify(resumeData, null, 2)}

        Requirements:
        1. Keep the exact same JSON structure
        2. Enhance content with stronger action verbs
        3. Add relevant keywords from job description
        4. Make achievements more quantifiable
        5. Optimize for ATS compatibility
        6. Return ONLY the enhanced JSON object`
      : `You are an expert resume writer. Enhance this resume to be more impactful.
        Return ONLY a valid JSON object with the enhanced content, keeping the exact same structure.
        Do not include any other text or markdown formatting in your response.
        
        Resume Data:
        ${JSON.stringify(resumeData, null, 2)}

        Requirements:
        1. Keep the exact same JSON structure
        2. Use stronger action verbs
        3. Make achievements more quantifiable
        4. Improve professional language
        5. Optimize for ATS compatibility
        6. Return ONLY the enhanced JSON object`

    console.log("\n=== Enhancing Resume with Gemini ===")
    console.log("Sending prompt to Gemini:")
    console.log(prompt)
    console.log("===================================\n")

    const result = await model.generateContent(prompt)
    const response = await result.response
    
    try {
      const text = response.text()
      if (!text) throw new Error("Empty response from AI")
      
      // First try direct JSON parse
      try {
        return JSON.parse(text)
      } catch {
        // If direct parse fails, try to extract JSON
        const jsonMatch = text.match(/\{[\s\S]*\}/)
        if (jsonMatch) {
          return JSON.parse(jsonMatch[0])
        }
        throw new Error("No valid JSON found in response")
      }
    } catch (parseError) {
      console.error("JSON parsing error:", parseError)
      throw new Error("Failed to parse enhanced resume data")
    }

    throw new Error("Failed to parse enhanced JSON from AI response")
  } catch (error) {
    console.error("Error enhancing resume with AI:", error)
    throw error
  }
}

export async function calculateATSScore(resumeData: any, jobDescription: string) {
  try {
    const model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" })

    const prompt = `You are an ATS system analyzer. Calculate the compatibility score between this resume and job description.
    Return ONLY a valid JSON object with this exact structure, no other text or formatting:
    {
      "score": 85,
      "suggestions": ["suggestion1", "suggestion2", "suggestion3"],
      "matchedKeywords": ["keyword1", "keyword2"],
      "missingKeywords": ["missing1", "missing2"],
      "analysis": {
        "keywordMatch": 80,
        "formatScore": 90,
        "skillsMatch": 85,
        "experienceMatch": 80
      }
    }

    Job Description:
    ${jobDescription}
    
    Resume Data:
    ${JSON.stringify(resumeData, null, 2)}

    Requirements:
    1. Use the exact JSON structure shown above
    2. Score should be 0-100 based on overall match
    3. Identify key missing skills or keywords
    4. Provide actionable suggestions
    5. Return ONLY the JSON object with scores and analysis`

    console.log("\n=== Calculating ATS Score with Gemini ===")
    console.log("Sending prompt to Gemini:")
    console.log(prompt)
    console.log("======================================\n")

    const result = await model.generateContent(prompt)
    const response = await result.response
    
    try {
      const text = response.text()
      if (!text) throw new Error("Empty response from AI")
      
      // First try direct JSON parse
      try {
        return JSON.parse(text)
      } catch {
        // If direct parse fails, try to extract JSON
        const jsonMatch = text.match(/\{[\s\S]*\}/)
        if (jsonMatch) {
          return JSON.parse(jsonMatch[0])
        }
        throw new Error("No valid JSON found in response")
      }
    } catch (parseError) {
      console.error("JSON parsing error:", parseError)
      throw new Error("Failed to parse ATS analysis data")
    }

    throw new Error("Failed to parse ATS analysis from AI response")
  } catch (error) {
    console.error("Error calculating ATS score:", error)
    throw error
  }
}

export async function scrapeJobDescription(url: string) {
  try {
    // This would typically be done on the backend with Puppeteer
    // For now, we'll return a placeholder
    const response = await fetch("/api/scrape-job", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ url }),
    })

    if (!response.ok) {
      throw new Error("Failed to scrape job description")
    }

    const data = await response.json()
    return data.jobDescription
  } catch (error) {
    console.error("Error scraping job description:", error)
    throw error
  }
}
